### PR TITLE
Handle redirects in the pipeline

### LINKED ISSUE #22746: Handle redirects as part of the pipeline
HMAC policy needs to be able to react to new URI.

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/Azure/azure-sdk-for-net/issues/22746

### CODE REVIEW

**Review by christothes — APPROVED** [review]

File: sdk/core/Azure.Core/src/Pipeline/HttpClientTransport.cs [hunk 1]
```diff
@@ -153,9 +153,15 @@ private static HttpMessageHandler CreateDefaultHandler()
...
+            return new SocketsHttpHandler()
+            {
+                AllowAutoRedirect = false
+            };
 #else
-            return new HttpClientHandler();
+            return new HttpClientHandler()
+            {
+                AllowAutoRedirect = false
+            };
```
[the remark this answers is not in the document]
Comment (pakrym) [reply sdk/core/Azure.Core/src/Pipeline/HttpClientTransport.cs:164]: That's evil :)

File: sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+                    break;
+                }
+
+                response.Dispose();
+
+                // Clear the authorization header.
+                request.Headers.Remove(HttpHeader.Names.Authorization);
+
+                if (AzureCoreEventSource.Singleton.IsEnabled())
```
[the remark this answers is not in the document]
Comment (pakrym) [reply sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs:59]: To avoid converting Uris to strings if the ES is disabled.

**Comment by chrwhit** [comment]
After this update, I assume I should remove my changes in Communications, correct?

**Review by chrwhit — APPROVED** [review]

**Comment by pakrym** [comment]
>After this update, I assume I should remove my changes in Communications, correct?

Correct. You would also need to add a ProjectReference to Azure.Core (to get the new feature) and flip it back to PackageReference when Azure.Core ships with the fix.

**Comment by chrwhit** [comment]
> > After this update, I assume I should remove my changes in Communications, correct?
> 
> Correct. You would also need to add a ProjectReference to Azure.Core (to get the new feature) and flip it back to PackageReference when Azure.Core ships with the fix.

Can we just wait for the package?

**Comment by pakrym** [comment]
>Can we just wait for the package?

That works too!

File: sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                        AzureCoreEventSource.Singleton.RequestRedirectCountExceeded(request.ClientRequestId, request.Uri.ToString(), redirectUri.ToString());
+                    }
+
+                    break;
+                }
+
+                response.Dispose();
+
+                // Clear the authorization header.
+                request.Headers.Remove(HttpHeader.Names.Authorization);
```
Comment (pharring) [inline sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs:57]: Why?

Comment (pakrym) [reply sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs:57]: https://github.com/dotnet/runtime/issues/26475#issuecomment-432878482

TLDR to prevent accidental secret leaks.

Comment (pharring) [reply sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs:57]: Thanks.
